### PR TITLE
Fix potential data inconsistency under heavy ddl operation

### LINKED ISSUE #5032: Data is inconsistent between TiFlash and TiKV after changing column types
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run the full-stack test: https://github.com/pingcap/tiflash/blob/335b124cab124828aadc318b7355f8390950f032/tests/fullstack-test2/ddl/alter_column_when_pk_is_handle.test, it will fail probability: https://ci.pingcap.net/blue/organizations/jenkins/tiflash-ghpr-integration-tests/detail/tiflash-ghpr-integration-tests/6648/pipeline

### 2. What did you expect to see? (Required)
Pass the test.
### 3. What did you see instead (Required)
The test fail, 
飞书20220531-000239
Data is inconsistent between TiFlash and TiKV, the log shows that TiFlash use default value instead of `0`:
image
The `col_6` does not exist in stable:
image

### 4. What is your TiFlash version? (Required)
7dff46338e179a047127abb151d2db586a5b9cce

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: close #5032 

Problem Summary: Currently we cache decoding schema for decoding raft data if a table schema doesn't change. And we judge whether a table schema has changed based on the table schema version.

But the schema version is not strictly consistent with the actual table schema which can be seen at https://github.com/pingcap/tiflash/blob/master/dbms/src/TiDB/Schema/SchemaBuilder.cpp#L362. That is when applying different schema changes in a diff, the table schema version will be set to the latest schema version after the first schema change is applied.

More concretely, when a lossy ddl change occurs, it will trigger `drop column` and `add column` schema changes and also rewrite some data at the same time.  The schema changes will be applied one by one, and because tiflash updates the table schema version ahead of time when applying schema diff, after applying the `drop column` schema change, it will update the schema version to the latest schema version.

And if the decode thread tries to obtain the current schema for decoding data before the `add column` is applied, the current schema and the latest schema version will be cached. Then after the subsequent `add column` operation is applied, the table schema version will not be updated, so the cache of the decode thread will not be invalidated. 

Therefore, the decode thread will decode the new data with an older schema, considering that the new added column is a dropped column and discarding it.

In addition, there is a lower chance of triggering this problem in the case of frequent add column and insert data.

### What is changed and how it works?
Add an internal schema version for `DecodingStorageSchemaSnapshot`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix potential data inconsistency under heavy ddl operation
```

### CODE REVIEW

File: dbms/src/Storages/StorageDeltaMerge.h [hunk 1]
```diff
@@ -238,6 +238,11 @@ class StorageDeltaMerge
 
     mutable std::mutex decode_schema_mutex;
     DecodingStorageSchemaSnapshotPtr decoding_schema_snapshot;
+    // The following two members must be used under the protection of table structure lock
+    bool decoding_schema_changed = false;
+    // internal version for `decoding_schema_snapshot`
+    size_t decoding_schema_version = 1;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/StorageDeltaMerge.h:244]: We use `Int64` as the param type in `void releaseDecodingBlock(Int64 block_decoding_schema_version, ...`, should also use Int64 here?

Comment (lidezhu) [reply dbms/src/Storages/StorageDeltaMerge.h]: Fixed.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by flowbehappy — APPROVED** [review]

**Comment by lidezhu** [comment]
> In response to a cherrypick label: failed to apply #5044 on top of branch "release-6.1":
> 
> ```
> failed to git commit: exit status 1
> ```

release-6.1 has been manually cherry-picked in this pr: https://github.com/pingcap/tiflash/pull/5046

**Pull request merged**